---
title: "Post-Impute_RSV"
author: "Mari Johnson"
date: "05/04/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, root.dir = 'home/mari/RSV/post-imp')
```

### Concatenation
```{bash}
bash concat.sh
```
### QC genotyped SNPs for frequency

```{bash PLINK set up and QC}

#Convert VCF to plink2 pfile

plink2 --vcf RSV_imputed.vcf.gz --geno 0.1 --mind 0.1 --make-pgen --out RSV_imp_QC
plink2 --pfile RSV_imp_QC --geno 0.05 --mind 0.05 --maf 0.01 --make-pgen --out RSV_imp_QC
#SNPs missing from 5% of population MAF <5 %
#So out of 340 the SNPs would have to be in 12 people
#10% cut off is probably better for assoc analysis in 34 people

#Convert to bfile
#Plink2 pmerge function is still under development

plink2 --pfile RSV_imp_QC --geno 0.05 --mind 0.05 --maf 0.01 --make-bed --out RSV_imp_QC

```
**RSV post-imputation results:**
- No varients removed after first QC round
- 25531828 variants removed due to allele frequency threshold(s)
- 8,917,131 variants remaining after main filters

```{bash Run Ethnicity Script}
#Script to check ethnicity data against 1000genomes
#liftover back to 37 is required unfortunately
bash Ethnicity_Check.sh
# Ref merging errors - come back to reformatting see test4/3.sh and 1000g and ref_cleaning scripts
```

```{bash FIlter IDs}
plink2 --pfile RSV_imp_QC --keep study_samples.txt --make-pgen --out RSV_Oxford
#318 samples
```

### PCA

- Use a PCA to account for population structure in data
- Check ancestry first by plotting PCA and colour coding by ethnicity
- Can also merge with public population databases

- Include PCA's as covariates in assocation model

```{bash Run PCA}
#plink2 --pfile RSV_imp_QC --pca --out all_Samples
plink2 --pfile RSV_Oxford --pca 
```

```{r Set up PCA in R, eval=TRUE}
pca <- fread("plink2.eigenvec")
#Remove hashtag IID in txt file

library("ggplot2")

ggplot(pca, aes(PC1,PC2)) + geom_point()
eigenval <- fread("plink2.eigenval")

#Colour by ethnicity
covar <- fread("covar_all.txt") 

pca_covar <- left_join(pca, covar, by = "IID")
#matched 318 rows :)

pca_covar$Sex <- as.factor(pca_covar$Sex)
levels(pca_covar$Sex) <- c("Male", "Female")

#Plot PCA coloured by Sex
pca_covar %>% ggplot(aes(PC1,PC2, color=Sex)) + geom_point()

#Calculate Percentage variance explained
pve <- data.frame(PC = 1:10, pve = round(eigenval/sum(eigenval)*100,2))

colnames(pve) <- c("PC", "pve")
pve$PC <- as.factor(pve$PC)

p <- pve %>% ggplot(aes(x=PC,y=pve)) + geom_col()

pve %>% ggplot(aes(x=PC,y=pve, fill = PC)) + geom_line()
p + labs(x = "Principle Component", y = "Percentage Varience Explained")
#Elbow at PC4 #Use first 4 PCAs as covars 

#Make new covar file to use
write.table(pca_covar, file = "pca_covar.txt", row.names = F, quote = F, sep = "\t")
```

### Conclusions

- Include first 4 PC's as covariates in association analysis

### Association Analysis

Initial linear association model (GLM) with --pheno-name RESV_total_score
Covars to include = PC1-4, Sex, has_pre_ex_con, baseline_age_at_visit
https://www.cog-genomics.org/plink/2.0/assoc

```{bash Association Analysis}
#mkdir -p Assoc
plink2 --pfile RSV_Oxford --pheno pheno.txt --pheno-name RESV_total_score --covar covar.txt --covar-name Sex, PC1, PC2, PC3, PC4, has_pre_ex_con, baseline_age_at_visit --covar-variance-standardize --glm --out Assoc/
```

```{r Plot Assoc Results, eval = TRUE}

gwas <- fread("Assoc/RESV_total_score.glm.linear")

#rename cols for qqman
colnames(gwas) <- c("CHR", "BP", "SNP", "REF", "ALT", "A1","TEST","OBS_CT","BETA",        
                           "SE","T_STAT","P","ERRCODE")

gwas$TEST <- as.factor(gwas$TEST)
gwas<- filter(gwas, TEST == "ADD") #Keep just genetic testing

#Plot Manhattan
library(qqman)
manhattan(gwas, ylim = c(0, 8), cex = 0.6, cex.axis = 0.9, genomewideline = F, col = c("mediumseagreen", "darkolivegreen4"))

#Inspect top snps:
tophits <- filter(gwas, P < 10e-5)
head(tophits)
#Write top hits file 
write.table(tophits, file = "Assoc/tophits.txt", quote = F, row.names= F, sep = "\t")
```


