---
title: "Genetic Associations of RSV Severity"
author: "Mari Johnson"
date: "2022-09-20"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = TRUE, warning = FALSE, message = FALSE)
knitr::opts_knit$set(root.dir = '~/RSV/data/post-imp/Assoc')
```
### Aims:

- Using imputed and cleaned data, associate genome-wide SNP frequency in our cohort with RSV severity (GWAS)
- Association analysis can be modified to also look at regions of interest (CGAS)
- Check Q-Q plot and genomic inflation values
- Visualise GWAS significance with manhattan plots
- Filter and format top  hit data for downstream annotation


### Association Analysis

-   General linear association model (GLM): pheno-name RESV_total_score

-   Covariates to include: PC1-4, Sex, has_pre_ex_con, baseline_age_at_visit\
    <https://www.cog-genomics.org/plink/2.0/assoc>

```{bash Association Analysis}
#mkdir -p Assoc
plink2 --bfile RSV_Oxford_nosib --pheno pheno2.txt --pheno-name RESV_total_score --covar pca_covar2.txt --covar-name Sex, PC1, PC2, PC3, PC4, has_pre_ex_con, baseline_age_at_visit --covar-variance-standardize --glm --out nosib2
```

```{bash Case/Control Analysis}
plink2 --bfile RSV_Oxford_nosib --pheno pheno2.txt --pheno-name RSV  --covar pca_covar2.txt --covar-name Sex, PC1, PC2, PC3, PC4, has_pre_ex_con, baseline_age_at_visit --covar-variance-standardize --glm --out nosib
```

### Linear Regression

Plink2 fits generalised linear mixed effects model:

y = GùõΩ + CùõΩ + e

-   Where, y is our response variable - ReSVinet score

-   G is our genotype dosage matrix

-   C is our covariate matrix

-   e is the error standard error (or residual) of our model estimate

-   ùõΩ are our model coefficients of determination. So, for each SNP we are modelling whether zero, one or two copies of an allele in our population, against our linear y variable, RSV severity.

-   Then test to see if there is a relationship, how strong the effect is, if it is signifcant

![](images/Taipei2015session02_page-0012.jpg)

The null hypothesis is that there is no linear relationship. Therefore if;\
ùõÉ = 0 the allele shows no relationship with our phenotype score\
ùõÉ \> 0 the allele is associated with a more positive score\
ùõÉ \< 0 the allele is associated with a more negative score

The significance of this relationship is then tested by calculating the T statistic:\
\
**t = b / SE**

-   **b** is the estimated value of linear slope or the coefficient of the predictor variable.

-   **SE~*b*~** represents the standard error of the coefficient estimation which can be estimated using the following formula:

    -   **SE = S / ‚àöN** Where S represents the standard deviation and N represents the total number of data points

If the p-value that corresponds to t is less than some threshold (e.g. Œ± = .05); then we reject the null hypothesis and conclude that there is a statistically significant relationship between the predictor variable and the response variable

```{r Plot Assoc Results, eval = TRUE, cache=TRUE}


# RESV severity Assoc -----------------------------------------------------------
plot_dir <- c("~/RSV/data/post-imp/Assoc/plots/")
gwas2 <- fread("nosib2.RESV_total_score.glm.linear")

# Rename cols for qqman
colnames(gwas2) <- c("CHR", "BP", "SNP", "REF", "ALT", "A1","TEST","OBS_CT","BETA",        
                           "SE","T_STAT","P","ERRCODE")

gwas2 <-
  filter(gwas2, TEST == "ADD")#Keep just genetic testing


# top <- filter(gwas2, P <= 10e-5)

test <- gwas2 %>% filter(P <= 10e-2) %>% select(c("CHR", "BP", "SNP", "REF", "ALT", "P"))

test$SNP <- str_replace_all(test$SNP, ":", "_")


# write.table(top, file = "gwas2.txt", quote = F, row.names= F, sep = "\t")

write.table(test, file = "test.txt", quote = F, row.names= F, sep = "\t")


# Plot Manhattan
manhattan(gwas2, ylim = c(0, 8), cex = 0.6, cex.axis = 0.9, genomewideline = F, col = c("mediumseagreen", "darkolivegreen4"))

pdf(file =(paste(plot_dir,"manhattan_RESV2.pdf",sep ="")), paper = "a4r")
manhattan(gwas, ylim = c(0, 8), cex = 0.6, cex.axis = 0.9, genomewideline = F, col = c("mediumseagreen", "darkolivegreen4"))
dev.off()

# Inspect top snps: ------------------------------------------------------------
tophits <- filter(gwas, P < 10e-5)
head(tophits)
#Write top hits file 
write.table(tophits, file = "tophits2.txt", quote = F, row.names= F, sep = "\t")

tophits2 <- filter(gwas2, P < 10e-5)
write.table(tophits2, file = "tophits22.txt", quote = F, row.names= F, sep = "\t")


# Case control -----------------------------------------------------------------

case <- fread("nosib.RSV.glm.logistic.hybrid")
names(case) <- c("CHR", "BP", "SNP", "REF", "ALT", "A1", "FIRTH","TEST","OBS_CT","OR",        
                           "LOG(OR)_SE","Z_STAT","P","ERRCODE")

case <- filter(case, TEST == "ADD") 
case_tophits <- filter(case, P < 10e-5)

# pdf(file =(paste(plot_dir,"manhattan_RESV.pdf",sep ="")))
manhattan(case, ylim = c(0, 8), cex = 0.6, cex.axis = 0.9, genomewideline = F, col = c("mediumseagreen", "darkolivegreen4"))
# Much less sig, and is not suited to reflect a linear response to infection

```

```{bash CD14 association}
plink2 --bfile RSV_Oxford_nosib --chr 5 --from-bp 140630856 --to-bp 140634071 --pheno pheno2.txt --pheno-name RESV_total_score --covar pca_covar2.txt --covar-name Sex, PC1, PC2, PC3, PC4, has_pre_ex_con, baseline_age_at_visit --covar-variance-standardize --glm --out CD14
```

```{r CD14 results}

CD14 <- fread("CD14.RESV_total_score.glm.linear")
CD14 <- filter(CD14, TEST == "ADD")
# No sig results

```

```{bash}
# Prep for annotation
# cut in bash

cut -f 3 tophits22.txt > explore_hits.txt
sed -i 's/chr//g' explore_hits.txt
cut -d ':' -f 1,2 explore_hits.txt > explore_hits2.txt

# Updates - server did not work
```

```{r gtex}

gtex <- read.csv(file = "gtex.csv")
gtex <- drop_na(gtex) %>% filter(dbSNP != "none")
gtex$tissue <- rep("Lung", length(gtex$dbSNP))
gtex <- gtex[1:400,]
write.csv(gtex, file = "gtex_lung.csv", row.names = F, quote = F)

```