---
title: "Association_testing"
output: html_notebook
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = TRUE, warning = FALSE, message = FALSE)
knitr::opts_knit$set(root.dir = '~/RSV/data/post-imp/Assoc')
setwd("~/RSV/data/post-imp/Assoc")
```

### Association Analysis

-   Initial linear association model (GLM) with --pheno-name RESV_total_score

-   Covariates to include = PC1-4, Sex, has_pre_ex_con, baseline_age_at_visit\
    <https://www.cog-genomics.org/plink/2.0/assoc>

-   No LD pruning yet

```{bash Association Analysis}
#mkdir -p Assoc
plink2 --pfile RSV_Oxford --pheno pheno.txt --pheno-name RESV_total_score --covar covar.txt --covar-name Sex, PC1, PC2, PC3, PC4, has_pre_ex_con, baseline_age_at_visit --covar-variance-standardize --glm --out Assoc/
```

```{r Plot Assoc Results, eval = TRUE}

gwas <- fread("Assoc/RESV_total_score.glm.linear")

#rename cols for qqman
colnames(gwas) <- c("CHR", "BP", "SNP", "REF", "ALT", "A1","TEST","OBS_CT","BETA",        
                           "SE","T_STAT","P","ERRCODE")

gwas$TEST <- as.factor(gwas$TEST)
gwas<- filter(gwas, TEST == "ADD") #Keep just genetic testing

#Plot Manhattan
library(qqman)
manhattan(gwas, ylim = c(0, 8), cex = 0.6, cex.axis = 0.9, genomewideline = F, col = c("mediumseagreen", "darkolivegreen4"))

#Inspect top snps:
tophits <- filter(gwas, P < 10e-5)
head(tophits)
#Write top hits file 
write.table(tophits, file = "Assoc/tophits.txt", quote = F, row.names= F, sep = "\t")
```

To do:

-   Label Manhattan plot with top hit genes

-   SNPs of interest plot bar charts

-   Summarise nexus data intronic etc. with pie chart

-   SNP nexus interactome

-   Add in updated PCA plots

### Plots

    ```{r explore tophits}

    top <- read.csv(file = "/nexus/gwass_summary_data.csv")
    getwd()
    # Calculate frequency of categories

    levels(top$Predicted_Function)
    levels(top$Feature_Type)
    levels(top$Epigenome_Feature)
    levels(top$Epigenome_Feature)
    levels(top$Symbol)

    freq <- 
      setDT(top) %>%
        top[ , .N, keyby = feature] 
    # Pie chart of general feature type

    ggpie(top, "value", label = labs,
       lab.pos = "in", lab.font = "white",
       fill = "predicted_function", color = "white",
       palette = c("#00AFBB", "#E7B800", "#FC4E07"))



    # Pie chart of histones

    # Manhattan
    # filter top genes per function take top sig, and use that as the label

    topM <- 
      top %>%
      group_by(symbol) %>%
      
      apply(group, MARGIN = 1,2, function(x){
        
      })
      
    ```

```{r plots}

top <- read.csv(file = "/nexus/gwass_summary_data.csv")
getwd()

```
